File: codes/hash_table/202.happy-number/happy-number.go
package main

func isHappy(n int) bool {
	history := make(map[int]bool)
	tmp := 0
	getTmp := func() {
		for n != 0 {
			t := n % 10
			tmp += t * t
			n /= 10
		}
	}
	for {
		getTmp()
		if tmp == 1 {
			return true
		} else {
			if history[tmp] {
				return false
			} else {
				history[tmp] = true
				n = tmp
				tmp = 0
			}
		}
	}
}

File: codes/hash_table/202.happy-number/happy-number.cpp
#include <unordered_set>
using namespace std;
class Solution {
public:
    int getDigit(int &n) {
        int t = n % 10;
        n = n / 10;
        return t * t;
    }
    bool isHappy(int n) {
        unordered_set<int> history;
        while (true) {
            int tmp = 0;
            while (n) {
                tmp += getDigit(n);
            }
            if(tmp==1){
                return true;
            }else{
                auto ret = history.insert(tmp);
                if(ret.second==false){
                    return false;
                }else{
                    n = tmp;
                }
            }
        }
    }
};